module/bdev/aio/bdev_aio_sync.c

```diff
@@ -0,0 +1,235 @@
+#include "bdev_aio.h"
+#include "spdk/env.h"
+#include "spdk/thread.h"
+#include "spdk/bdev_module.h"
+
+#include <libaio.h>
+
+#include "bdev_aio_task.h"
+#include "bdev_aio_sync.h"
+
+#define MAX_QUEUE_LEN 1024
+
+struct queue
+{
+	void *message[MAX_QUEUE_LEN];
+	size_t used;
+};
+
+static struct queue g_queue;
+
+static void
+_init_queue(void)
+{
+	g_queue.used = 0;
+}
+
+static int
+_enqueue(void *message)
+{
+	if (g_queue.used < MAX_QUEUE_LEN)
+	{
+		g_queue.message[g_queue.used++] = message;
+
+		return 0;
+	}
+
+	return -1;
+}
+
+static size_t
+_dequeue(void *message[], size_t n)
+{
+	size_t i;
+
+	if (g_queue.used > n)
+	{
+		for (i = 0; i < n; i++)
+		{
+			message[i] = g_queue.message[i];
+		}
+
+		g_queue.used -= n;
+
+		for (i = 0; i < g_queue.used; i++)
+		{
+			g_queue.message[i] = g_queue.message[n + i];
+		}
+
+		return n;
+	}
+
+	if (g_queue.used > 0)
+	{
+		n = g_queue.used;
+		g_queue.used = 0;
+
+		for (i = 0; i < n; i++)
+		{
+			message[i] = g_queue.message[i];
+		}
+
+		return n;
+	}
+
+	return 0;
+}
+
+static pthread_mutex_t g_mutex;
+
+static pthread_cond_t g_cond;
+
+void
+aio_complete(struct spdk_bdev_io *bdev_io, int status)
+{
+	spdk_bdev_io_complete(bdev_io, (status == 0) ? SPDK_BDEV_IO_STATUS_SUCCESS : SPDK_BDEV_IO_STATUS_FAILED);
+}
+
+static void
+aio_call_complete_fn(void *arg)
```

Comment:
I'd rather have these functions called something more descriptive perhaps. What about

aio_sync_unmap_request and _complete ?

Reply:
maybe aio_sync_{request,complete} as, strictly speaking, they're not used exclusively for unmap.